File: core/src/main/scala/dqsuite/DQSuiteDatasetContext.scala
package dqsuite
import com.amazon.deequ.VerificationResult
import com.amazon.deequ.repository.{MetricsRepository, ResultKey}
import com.amazon.deequ.suggestions.ConstraintSuggestionResult
import dqsuite.config.SourceConfig
import dqsuite.runners.{PostprocessRunner, ProfilingRunner, SchemaCheckResult, SchemaCheckRunner, ValidationRunner}
import org.apache.spark.sql.DataFrame

import java.net.URI

case class DQSuiteDatasetContext(
  suiteContext: DQSuiteContext,
  config: SourceConfig,
  metricsPath: URI,
  resultPath: URI,
  resultKey: ResultKey,
  repository: MetricsRepository
) {
  def profile(
    df: DataFrame
  ): ConstraintSuggestionResult = {
    ProfilingRunner(this).run(df)
  }

  def checkSchema(
    df: DataFrame,
    emptyStringAsNull: Boolean = true
  ): SchemaCheckResult = {
    SchemaCheckRunner(this, emptyStringAsNull = emptyStringAsNull).run(df)
  }

  def validate(
    df: DataFrame,
    anomalyDetection: Boolean = true
  ): VerificationResult = {
    ValidationRunner(this, anomalyDetection).run(df)
  }

  def postprocess(
    df: DataFrame
  ): DataFrame = {
    PostprocessRunner(this).run(df)
  }
}

File: python/dqsuite/dqsuite_dataset_context.py
from typing import Any

from pydeequ.verification import VerificationResult
from pyspark.sql import DataFrame, SQLContext

from dqsuite.schema_check_result import SchemaCheckResult
from dqsuite.utils import PY4JClassWrapper


class DQSuiteDatasetContext(PY4JClassWrapper):
    def profile(self, df) -> Any:
        return self._callj("profile", df._jdf)

    def checkSchema(self, df, emptyStringAsNull: bool = True) -> SchemaCheckResult:
        return SchemaCheckResult(self._spark_session, self._callj("checkSchema", df._jdf, emptyStringAsNull))

    def validate(self, df, anomalyDetection: bool = True) -> VerificationResult:
        return VerificationResult(self._spark_session, self._callj("validate", df._jdf, anomalyDetection))

    def postprocess(self, df) -> DataFrame:
        dfj = self._callj("postprocess", df._jdf)

        sql_ctx = SQLContext(
            sparkContext=self._spark_session._sc,
            sparkSession=self._spark_session,
            jsqlContext=self._spark_session._jsparkSession.sqlContext(),
        )
        return DataFrame(dfj, sql_ctx)
